quinn-proto/src/bytes_source.rs

```diff
@@ -0,0 +1,231 @@
+//! Defines types which can be lazily converted into `Bytes` chunks
+
+use bytes::Bytes;
+
+/// A source of one or more buffers which can be converted into `Bytes` buffers on demand
+///
+/// The purpose of this data type is to defer conversion as long as possible,
+/// so that no heap allocation is required in case no data is writable.
+pub trait BytesSource {
+    // Returns the total amount of bytes to write
+    fn len(&self) -> usize;
+    /// Limits the source to the new size
+    fn limit(&mut self, new_len: usize);
+    /// The amount of bytes and chunks consumed from this source
+    fn consumed(&self) -> Written;
```

Comment:
Is there a benefit to this being tracked by the source? Intuitively it seems like the accumulators could easily be maintained source-independently in the ultimate `fn write` implementation.

Reply:
Kind of the same answer as for limits. It more or less requires all intermediate code to track the same information that the source already tracks. And we can't drop the information from there, since it also requires it to determine the offsets of the next data to write.

Reply:
I'd think there's even less communication between intermediate layers here than in the `limit` case. We have exactly one location that ever calls `pop_chunk`, and it can very naturally return the `Written` counters back up the stack.

Reply:
Here is an attempt: https://github.com/quinn-rs/quinn/compare/main...Matthias247:bytes_write_2?expand=1

Ignore most of the `ByteSource` implementation. It's not updated and broken, just has the updated API.
Most things seem to work fine. However the loop in `Send::write` is broken: It can't properly populate `Written::chunks`, since it doesn't know whether `pop_chunk` returned a full chunk or whether it was trunated due to the limit.

In order to fix this, we we would need to add either a bool to `pop_chunk` which indicates if this consumed a full buffer (which is however somewhat weird, since the internal buffer structures are transparent to the caller), or return a `Option<Written, Bytes>` from `pop_chunk`.

Reply:
So returning `Option<(Written, Bytes)>` wasn't optimal either, because it didn't allow to indicate we consumed chunks in the source if no `Bytes` elements had been pulled (which can happen in case empty `Bytes` elements are enqueued). It's obviously a borderline thing of invalid input, but I still think we need to handle it correctly.

One option to fix it is to change the return value to `(Option<Bytes>, Written)` or `(Option<Bytes>, usize)`, so that consumed chunks can be reported independent of pulling a chunk. I updated for this in the last version of the linked other variant.

I'm split on whether I like it better than this approach here. Yes - the trait is smaller. However returning a consumed chunks number is weird, since the trait nowhere else mentions multiple chunks.

Reply:
The new branch actually looks pretty good to me in its current state. The simpler interface makes the semantics a lot more obvious, and only mentioning multiple chunks in one place isn't obscure when there's only one method on the trait to begin with. I agree with the decision re: return values, as well, except that I wonder if `Option<Bytes>` could be simplified to `Bytes` by returning an empty `Bytes` when appropriate.

Reply:
ok. I will then update to it. I had the same thought about `Bytes::new()`. It might be actually be a bit more efficient. But on the other hand the explicit `None` is a good signal to stop iteration. I have no strong thoughts on it.

Reply:
I now updated for `Bytes::new` since I think it looks a bit more concise on the consumer side.